### ISSUE TITLE
TRY004 applies when unrelated values are type-checked

### ISSUE THREAD

**Issue by dscorbett** [astral-sh/ruff#12287]
TRY004 assumes that if any exception is raised, subject to a type-checking `if` condition, the exception should be a `TypeError`. That is only true when an argument has the wrong type. If the exception happens for a different reason (e.g. invalid state) then `TypeError` isn’t necessarily appropriate (even if the invalid state is detected by checking the type of an instance variable).

[The Python documentation for `TypeError`](https://docs.python.org/3/library/exceptions.html#TypeError) says:
> Passing arguments of the wrong type (e.g. passing a [`list`](https://docs.python.org/3/library/stdtypes.html#list) when an [`int`](https://docs.python.org/3/library/functions.html#int) is expected) should result in a [`TypeError`](https://docs.python.org/3/library/exceptions.html#TypeError), but passing arguments with the wrong value (e.g. a number outside expected boundaries) should result in a [`ValueError`](https://docs.python.org/3/library/exceptions.html#ValueError).

Example code:
```python
class FreezableList:
    def __init__(self):
        self.delegate = []

    def freeze(self):
        self.delegate = tuple(self.delegate)

    def append(self, value):
        if isinstance(self._delegate, list):
            self.delegate.append(value)
        else:
            raise ValueError("Appending to a frozen list")
```

Current behavior:
```console
$ ruff --version
ruff 0.5.1
$ ruff check --isolated --select TRY004 --output-format concise try004_example.py
try004_example.py:12:13: TRY004 Prefer `TypeError` exception for invalid type
Found 1 error.
```

A heuristic could be that, if the value in the `isinstance` call is neither an argument to the method or function nor is derived from an argument, TRY004 should not report anything.

**Comment by dhruvmanila** [comment]
Thanks for opening this issue! Technically, I think the heuristic is correct to raise a violation in this case. I'm not sure whether raising a `ValueError` makes more sense here because the "value" is not the problem here. Personally, I would raise a `TypeError` here because the main issue is in the type itself. CPython also raises a `TypeError` in case you try to concatenate a frozenset:

```
In [1]: x = frozenset(['a', 'b', 'c'])

In [2]: x + frozenset(['d'])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[2], line 1
----> 1 x + frozenset(['d'])

TypeError: unsupported operand type(s) for +: 'frozenset' and 'frozenset'
```

> A heuristic could be that, if the value in the `isinstance` call is neither an argument to the method or function nor is derived from an argument, TRY004 should not report anything.

I'm not sure if that would solve your use-case because `self.delegate` would be considered as derived from the `self` argument. Or, am I misunderstanding what you're trying to say?

**Comment by dscorbett** [comment]
> Personally, I would raise a `TypeError` here because the main issue is in the type itself.

I raise `ValueError` because the fact that I’m using `isinstance` is an internal implementation detail. For example, here is an alternative implementation of that class:
```python
class FreezableList:
    def __init__(self):
        self._delegate = []
        self._frozen = False

    @property
    def delegate(self):
        return tuple(self._delegate) if self._frozen else self._delegate

    def freeze(self):
        self._frozen = True

    def append(self, value):
        if not self._frozen:
            self.delegate.append(value)
        else:
            raise ValueError("Appending to a frozen list")
```

Both versions of the class implement essentially the same API, so they should presumably raise the same errors under the same circumstances, regardless of whether `isinstance` is used internally. The error is fundamentally about calling a method when the object is in the wrong state, not about types, so `ValueError` is more appropriate than `TypeError`.

> I'm not sure if that would solve your use-case because `self.delegate` would be considered as derived from the `self` argument. Or, am I misunderstanding what you're trying to say?

That’s true; I meant arguments besides `self`. `self` is an exception because it is set apart syntactically when calling the method, so it doesn’t feel like an argument in the same way. My intuition is that if the type-checking problem is related to a non-`self` argument, it is probably the caller’s fault for passing in a value of the wrong type, which warrants a `TypeError`, whereas if the problem is related to `self` (or one of its attributes), it has a different cause and need not warrant being surfaced to the caller as a `TypeError`.

**Comment by dhruvmanila** [comment]
> That’s true; I meant arguments besides `self`. `self` is an exception because it is set apart syntactically when calling the method, so it doesn’t feel like an argument in the same way. My intuition is that if the type-checking problem is related to a non-`self` argument, it is probably the caller’s fault for passing in a value of the wrong type, which warrants a `TypeError`, whereas if the problem is related to `self` (or one of its attributes), it has a different cause and need not warrant being surfaced to the caller as a `TypeError`.

I see. I think it makes sense, I'm trying to think through if there's any case where this might yield false negatives but can't find any. I'd love to hear others opinion.

**Comment by wolkiewiczk** [comment]
> A heuristic could be that, if the value in the isinstance call is neither an argument to the method or function nor is derived from an argument, TRY004 should not report anything.

I don't think that TRY004 should report if isinstance check is performed on an object derived from an argument. See this example:

```python
class MyTree(IntervalTree):

    def __init__(self, intervals=None):
        intervals = intervals or []
        for iv in intervals:
            if not isinstance(iv.data, MyDataclass):
                raise ValueError(
                    f"All intervals in {type(self).__name__} must have the 'data' attribute set to "
                    f"the instance of {MyDataclass.__name__} subclass."
                )
        super().__init__(*args, **kwargs)
```

In this case TRY004 is reported, even though the type of my argument is correct. Even the types of objects inside the iterable are correct, they are just required to have a certain class set as an attribute so I think that the `ValueError` is the correct choice here.

**Comment by AlexWaygood** [comment]
> in Java, I would use `IllegalStateException`

(FWIW, in the Python context I think `RuntimeError` is the closest analogy.)

**Comment by alicederyn** [comment]
I agree with the OP, the docs for `TypeError` are pretty clear:

> Raised when an operation or function is applied to an object of inappropriate type.

I have a function that parses a YAML file and raises a parsing error if anything is incorrect, including an incorrect type. This should result in a single consistent subclass of `RuntimeError`, not a `TypeError`, because the parameter provided to the function is not of incorrect type (it is correctly a `Path`). Following the TRY004 recommendation would leak implementation details, and make it impossible for callers to distinguish errors in preconditions they can fix (passing in an incorrect type) from ones they cannot (a parsed file was incorrect).

**Comment by tsangwpx** [comment]
I have the following that are perfectly valid to not use `TypeError`.

```python
    @classmethod
    def from_public_key(cls, pem: bytes) -> Self:
        loaded = serialization.load_pem_public_key(pem)
        if not isinstance(loaded, rsa.RSAPublicKey):
            raise ValueError(f"not an RSA public key: {type(loaded).__name__}")  # noqa: TRY004
        ...
```

Also, with static type checkers, the case where `TypeError` should raise is minimal. this is because `TypeError`, to my understanding, raise if the argument violates the **documented** signature or usage. With type annotations and `typing.overload`, most violations can be reported by type checkers, not ruff.

the remaining cases should be untyped functions and the Any type. their usage is documented somewhere.
I suggest this rule should only apply to untyped code.
